### PR TITLE
[RNMobile] Social icons

### PR DESCRIPTION
**Pull request by lukewalczak**
## Description

~~**NOTE**: That PR should be merged after https://github.com/WordPress/gutenberg/pull/23299 since it's using the `LinkSettings` component coming from there.~~

~~**NOTE**: That PR should be merged after https://github.com/WordPress/gutenberg/pull/23465~~

That PR introduces a new block called `Social Icon`.

* Ref to gutenberg-mobile: https://github.com/wordpress-mobile/gutenberg-mobile/pull/2361

Testing:

* ios: https://github.com/wordpress-mobile/WordPress-iOS/pull/14259
* android: https://github.com/wordpress-mobile/WordPress-Android/pull/12124

## How has this been tested?

1. Open mobile app
2. Add `Social Icons` block

Expect: Social Icons is added with 4 icons, where WordPress icon is already fulfilled with the link and it's active _(has product color)_

3. Press inactive icon eg. `Facebook` icon
4. Press `link` icon in mobile toolbar above the keyboard
5. Fill `URL` eg. facebook.com and `Label`
6. Close sheet

Expect: Facebook social icon is active _(has product color)_

7. Deselect `Social Icons` 

Expect: Only active social icons left

8. Open post preview
9. Press the icon and check if forwarding to the link

## Screenshots 

* light mode

<details>

active | inactive
--- | ---
Screenshot 2020-06-05 at 16 32 44 | Screenshot 2020-06-05 at 16 33 10

</details>

* dark mode

<details>

active | inactive
--- | ---
Screenshot 2020-06-05 at 16 32 25 | Screenshot 2020-06-05 at 16 33 18

</details>

* action sheet

<details>

android | ios
--- | ---
Screenshot 2020-06-08 at 15 50 22 | Screenshot 2020-06-08 at 15 54 12

</details>

* URL sheet

<details>

android | ios
--- | ---
Screenshot 2020-06-09 at 10 29 57 | Screenshot 2020-06-09 at 10 27 34

</details>

## Types of changes

New feature: `Social Icons` block

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. 
- [x] My code follows the accessibility standards. 
- [x] My code has proper inline documentation. 
- [x] I've included developer documentation if appropriate. 
- [x] I've updated all React Native files affected by any refactorings/renamings in this PR.

### CODE REVIEW

File: packages/block-editor/src/components/block-mover/mover-description.native.js [hunk 1]
```diff
@@ -13,10 +13,8 @@ const horizontalMover = {
 	forwardButtonIcon: arrowRight,
 	backwardButtonHint: __( 'Double tap to move the block to the left' ),
 	forwardButtonHint: __( 'Double tap to move the block to the right' ),
-	/* translators: %s: block title e.g: "Paragraph" */
-	firstBlockTitle: __( 'Move %s Left' ),
-	/* translators: %s: block title e.g: "Paragraph" */
-	lastBlockTitle: __( 'Move %s Right' ),
+	firstBlockTitle: __( 'Move block left' ),
```
Comment (dratwas) [inline packages/block-editor/src/components/block-mover/mover-description.native.js:16]: It seems like something that should be done in a separate PR. WDYT? I'm rather a fan of small PRs w/o side tasks

File: packages/block-library/src/index.native.js [hunk 2]
```diff
@@ -130,20 +134,52 @@ const registerBlock = ( block ) => {
...
+ * Function to register a block variations e.g. social icons different types.
+ *
+ * @param {Object} block The block which variations will be registered.
+ *
+ */
+const registerBlockVariations = ( block ) => {
+	const { metadata, settings, name } = block;
+
+	settings.variations
+		.sort( ( a, b ) => {
```
Comment (dratwas) [inline packages/block-library/src/index.native.js:147]: Maybe instead of implementing own comparator, we could use `lodash.sortBy` with `title` as a key https://lodash.com/docs/#sortBy

```js
_.sortBy(settings.variations, 'title');
```

File: packages/block-editor/src/components/block-mover/mover-description.native.js (hunk 1 above)
Comment (lukewalczak) [reply packages/block-editor/src/components/block-mover/mover-description.native.js:16]: Ok, will move all the changes related to block mover to the separated PR ✌️

File: packages/block-library/src/index.native.js (hunk 2 above)
Comment (lukewalczak) [reply packages/block-library/src/index.native.js:147]: Hmm, makes sense (less code), I will refactor it then.

File: packages/block-library/src/index.native.js [hunk 3]
```diff
@@ -130,20 +134,52 @@ const registerBlock = ( block ) => {
...
 
-// Hide the Classic block
+// Hide the Classic block and SocialIcon block
 addFilter(
 	'blocks.registerBlockType',
 	'core/react-native-editor',
 	( settings, name ) => {
 		if (
-			name === 'core/freeform' &&
+			( name === 'core/freeform' || name === 'core/social-link' ) &&
```
Comment (dratwas) [inline packages/block-library/src/index.native.js:182]: Hmmm, do you think that there will be more of these cases in the future? Maybe it's worth to create an array with them and check if it includes that string?

```js
const hiddenBlocks = [core/freeform', 'core/social-link'];
if( hiddenBlocks.includes( name ) && hasBlockSupport( settings, 'inserter', true ) ) {
  ...
}
```

File: packages/block-library/src/social-link/edit.native.js [hunk 4]
```diff
@@ -0,0 +1,167 @@
...
+	const { url, service } = attributes;
+	const activeIcon =
+		styles[ `wp-social-link-${ service }` ] || styles[ `wp-social-link` ];
+
+	const inactiveIcon = usePreferredColorSchemeStyle(
+		styles.inactiveIcon,
+		styles.inactiveIconDark
+	);
+
+	const [ isLinkSheetVisible, setIsLinkSheetVisible ] = useState( false );
```
Comment (dratwas) [inline packages/block-library/src/social-link/edit.native.js:43]: WDYT about moving it at the top of the component? I think that it's more readable to have a declaration of state at the top than here.

File: packages/block-editor/src/store/selectors.js [hunk 5]
```diff
@@ -1325,7 +1325,7 @@ export const getInserterItems = createSelector(
 			return {
 				id,
 				name: blockType.name,
-				initialAttributes: {},
+				initialAttributes: blockType.initialAttributes || {},
```
Comment (lukewalczak) [inline packages/block-editor/src/store/selectors.js:1328]: @youknowriad would like to confirm it's not against the web and will not break anything.

File: packages/block-editor/src/components/block-list/index.native.js [hunk 6]
```diff
@@ -311,42 +312,55 @@ export class BlockList extends Component {
...
+			const selectedBlockClientId = getSelectedBlockClientId();
+
+			let blockClientIds = getBlockOrder( rootClientId );
+
+			// Display only block which fulfill the condition in passed `filterInnerBlocks` function
+			if ( filterInnerBlocks ) {
+				blockClientIds = filterInnerBlocks(
+					getBlockOrder( rootClientId )
+				);
+			}
```
Comment (dratwas) [inline packages/block-editor/src/components/block-list/index.native.js:341]: Hmm i believe we could refactor it a bit since we don't need to call `gerBlockOrder` twice if the `filterInnerBlocks` is true. Maybe we could do something like:
```js
const blockClientIds = filterInnerBlocks ? filterInnerBlocks(getBlockOrder( rootClientId )) : getBlockOrder( rootClientId );
```

File: packages/block-library/src/social-link/edit.native.js [hunk 7]
```diff
@@ -0,0 +1,188 @@
...
+			outputRange: [ inactiveIcon.color, activeIcon.color ],
+		} ),
+		stroke: '',
+	};
+
+	const { backgroundColor, color, stroke } = isValue
+		? activeIcon
+		: interpolationColors;
+
+	const options = {
```
Comment (dratwas) [inline packages/block-library/src/social-link/edit.native.js:86]: It seems like something that could be a const declared outside of the component.

File: packages/block-library/src/social-links/edit.native.js [hunk 8]
```diff
@@ -0,0 +1,142 @@
...
+		</View>
+	) );
+
+	const placeholderStyle = usePreferredColorSchemeStyle(
+		styles.placeholder,
+		styles.placeholderDark
+	);
+
+	function renderPlaceholder() {
+		return [ ...new Array( innerBlocks.length || 1 ) ].map( ( index ) => (
```
Comment (dratwas) [inline packages/block-library/src/social-links/edit.native.js:67]: We shouldn't use the index as a key but I imagine that in some cases this is impossible to avoid, however, I believe it should look like this :)
```js
.map( ( _, index ) => (
```

File: packages/block-library/src/social-link/edit.native.js [hunk 9]
```diff
@@ -0,0 +1,188 @@
...
+				<BlockControls>
+					<ToolbarGroup>
+						<ToolbarButton
+							title={ sprintf(
+								// translators: %s: social link name e.g: "Instagram".
+								__( 'Add link to %s' ),
+								socialLinkName
+							) }
+							icon={ link }
+							onClick={ () => setIsLinkSheetVisible( true ) }
```
Comment (dratwas) [inline packages/block-library/src/social-link/edit.native.js:153]: I'm wondering why you created the `onCloseSettingsSheet` where you call `setIsLinkSheetVisible` with false but you did not create a function for this usage :)

File: packages/block-editor/src/components/block-list/index.native.js [hunk 10]
```diff
@@ -311,42 +312,55 @@ export class BlockList extends Component {
...
+
+			const isReadOnly = getSettings().readOnly;
+
+			const rootBlockId = getBlockHierarchyRootClientId(
+				selectedBlockClientId
+			);
+			const hasRootInnerBlocks = !! getBlockCount( rootBlockId );
```
Comment (dratwas) [inline packages/block-editor/src/components/block-list/index.native.js:348]: I know it's not your change but could you please assign the `getBlockCount` to `blockCount` const and use it here and in the return of this select?

File: packages/block-library/src/social-link/edit.native.js (hunk 9 above)
Comment (lukewalczak) [reply packages/block-library/src/social-link/edit.native.js:153]: Looks like something which left after adjusting to new `LinkSettings` component, however, I will create an adequate function `onOpenSettingsSheet`.

File: packages/block-library/src/social-links/edit.native.js (hunk 8 above)
Comment (lukewalczak) [reply packages/block-library/src/social-links/edit.native.js:67]: I can agree that name `index` is misleading and maybe I should call it just `e` or `item` since it's only an array of numbers.

File: packages/editor/src/components/post-title/index.native.js [hunk 11]
```diff
@@ -26,13 +26,19 @@ import { pasteHandler } from '@wordpress/blocks';
 import styles from './style.scss';
 
 class PostTitle extends Component {
+	constructor( props ) {
```
Comment (dratwas) [inline packages/editor/src/components/post-title/index.native.js:29]: Changes in that file look like something that could be done in separate PR. I know that probably you fixed it during work on social links but i strongly recommend split PRs to smaller. Thanks to that it would be easier to review and test :)

Comment (lukewalczak) [reply packages/editor/src/components/post-title/index.native.js:29]: Moved to the separate PR: https://github.com/WordPress/gutenberg/pull/23683

File: packages/primitives/src/svg/index.native.js [hunk 12]
```diff
@@ -39,7 +49,7 @@ export const SVG = ( { className = '', isPressed, ...props } ) => {
 	const appliedProps = { ...props, style: styleValues };
 
 	return (
-		<Svg
+		<AnimatedSvg
```
Comment (geriux) [inline packages/primitives/src/svg/index.native.js:52]: I wonder if we should have all SVGs as animated components or just when we actually want to animate something. Do you think this would affect the performance in some way? By having all icons as Animated components?

File: packages/block-library/src/social-link/edit.native.js [hunk 13]
```diff
@@ -0,0 +1,192 @@
...
+
+const SocialLinkEdit = ( {
+	attributes,
+	setAttributes,
+	isSelected,
+	onFocus,
+} ) => {
+	const { url, service } = attributes;
+	const [ isLinkSheetVisible, setIsLinkSheetVisible ] = useState( false );
+	const [ isValue, setIsValue ] = useState( !! url );
```
Comment (geriux) [inline packages/block-library/src/social-link/edit.native.js:51]: It took me a minute to understand what this meant 😅, I'm usually not picky with names but what do you think if we change it to something like `hasUrl` or `isUrlSet` or something related to the `url`?

File: packages/block-editor/src/store/selectors.js (hunk 5 above)
Comment (geriux) [reply packages/block-editor/src/store/selectors.js:1327]: Yup, It'd be great to confirm that.

File: packages/primitives/src/svg/index.native.js (hunk 12 above)
Comment (lukewalczak) [reply packages/primitives/src/svg/index.native.js:52]: It's shouldn't affect, however for the sake of calm I will create a separate component for it. Thanks!

File: packages/block-library/src/social-link/edit.native.js (hunk 13 above)
Comment (lukewalczak) [reply packages/block-library/src/social-link/edit.native.js:51]: It makes sense!

File: packages/primitives/src/svg/index.native.js (hunk 12 above)
Comment (geriux) [reply packages/primitives/src/svg/index.native.js]: Awesome! Thanks!

File: packages/block-editor/src/store/selectors.js (hunk 5 above)
Comment (youknowriad) [reply packages/block-editor/src/store/selectors.js:1326]: Why would a block type define "initialAttributes"? I think that's the role of the block variations.

File: packages/block-library/src/social-link/edit.native.js [hunk 14]
```diff
@@ -193,4 +195,15 @@ const SocialLinkEdit = ( {
 	);
 };
 
-export default SocialLinkEdit;
+export default compose( [
+	withSelect( ( select, { clientId } ) => {
+		const { getBlock } = select( 'core/block-editor' );
+
+		const block = getBlock( clientId );
+		const name = block.name.substring( 17 );
```
Comment (geriux) [inline packages/block-library/src/social-link/edit.native.js:203]: While testing removing some social links I got a red screen related to `block.name` can we use `?.` or check if the block exists before accessing to `name` ?

**Review by geriux — APPROVED** [review]
LGTM! Nice work @lukewalczak! 👏 Tested it on the demo app (iOS/Android) and also on WordPress iOS.

File: packages/blocks/src/api/parser.js [hunk 15]
```diff
@@ -434,7 +434,8 @@ export function createBlockWithFallback( blockNode ) {
 	if ( name && name.indexOf( 'core/social-link-' ) === 0 ) {
 		// Capture `social-link-wordpress` into `{"service":"wordpress"}`
 		attributes.service = name.substring( 17 );
-		name = 'core/social-link';
+		// Display social link service name for mobile platform
+		name = Platform.OS === 'web' ? 'core/social-link' : name;
```
Comment (mcsf) [inline packages/blocks/src/api/parser.js:438]: Can you help me understand why the deprecated name needs to be kept in mobile?

My worries:
* Even though we should keep this compat shim for years to preserve older content, the idea is still to help move content forward by remapping block names, so that, when a user edits an old post, the Social Link markup is correctly updated.
* The shim is likely to move out of the parser and be placed somewhere more sustainable (e.g. external compat file). This could be provided by JS filters or some other form of hooking. Once we do that, it would be important that the shim _unconditionally_ reassigns `name` to `core/social-link`.

/cc @lukewalczak @hypest

Comment (lukewalczak) [reply packages/blocks/src/api/parser.js:438]: Hey @mcsf, that change was basically needed to display specific social icon name and its icon within toolbar on the mobile platform.
Without this change, it would be confusing as each `social-icon` would have the same label and the same icon after saving the post.

without parser change | with parser change
--- | ---
Screenshot 2020-07-30 at 14 44 08 | Screenshot 2020-07-30 at 14 45 01

The introduced change was the easiest solution to fulfil the requirement, but I understand your concerns and will double-check if we can achieve the same effect without parsers changes.

Comment (lukewalczak) [reply packages/blocks/src/api/parser.js:438]: However, could you please elaborate more on:

> The shim is likely to move out of the parser and be placed somewhere more sustainable (e.g. external compat file). This could be provided by JS filters or some other form of hooking. Once we do that, it would be important that the shim unconditionally reassigns name to core/social-link

Comment (mcsf) [reply packages/blocks/src/api/parser.js:438]: Thanks for the context.

It's clear then that support for Block Variations needs to be improved so that those interactions can take into account attributes like a variation's title or icon.

For comparison, even on the Web we let the variation show in certain places and let the base type show in others, which is also a problem but less confusing than in the apps:

Captura de ecrã 2020-07-30, às 14 12 11

See **open questions** about variations: https://github.com/WordPress/gutenberg/issues/16283#issuecomment-593404741

* * *

> However, could you please elaborate more on:
> > The shim is likely to move out of the parser […]

Sure. I'm referring to [this comment by Grzegorz](https://github.com/WordPress/gutenberg/pull/19887/files#r375181162).

* * *

Context:
* Block variations API: #16283, #19243 
* Social Icons implementation: #19887

Comment (lukewalczak) [reply packages/blocks/src/api/parser.js:438]: So I assume that the main goal is to revert that change and have equal HTML on both platforms?

Comment (mcsf) [reply packages/blocks/src/api/parser.js:438]: Yep, correct.

Comment (lukewalczak) [reply packages/blocks/src/api/parser.js:438]: I've created a [PR](https://github.com/WordPress/gutenberg/pull/24306) fixing that, please check it :v:
